docs/design/feature-attributes.md

```diff
@@ -0,0 +1,181 @@
+
+- Methods, fields, properties, and events which have a _feature requirement_ are guaranteed to be removed.
+
+- Methods which are in a _feature available_ scope but aren't entirely removed are guaranteed to have the method body replaced with a throwing instruction sequence.
+
+Thie latter can happen for methods in a type with _feature requirement_ (but that do not themselves have _feature requirement_) that are referenced outside of a _feature available_ scope. The reference to such a method may remain even though the type is never constructed. The callsite would produce a `NullReferenceException` and the method body is unreachable.
+
+## Alternatives
+
+One simplification would be to unify the concepts of _feature requirement_ with _feature available_, and treat both as similar to preprocessor symbols, where _any_ reference to a guarded type or member from an unguarded context is disallowed.
+
+The advantage of the specified model is that it allows some references without warning, giving some extra flexibility and making it easier to migrate existing code. The downside is that it might lead to preserving more code, whereas a simplified model could guarantee that all code related to a disabled feature is removed.
+
+Here is an example of a pattern which does not warn in the current model, but would warn with a simplified model. Assume that the code under `SomeFeatureIsSupported` is removed when the feature is unavailable.
+
+```csharp
+class FeatureConsumer {
+    static void Run() {
+        SomeFeatureProvider? some;
+        if (Features.SomeFeatureIsSupported)
+            some = new SomeFeatureProvider();
+        OtherFeatureProvider other = new();
+        Helper(some, other);
+    }
+
+    static void Helper(SomeFeatureProvider? some, OtherFeatureProvider other) {
+        some?.Use(); // This callsite would warn with the simplified model.
+        other.Use();
+    }
+}
+
+[RequiresSomeFeature]
+class SomeFeatureProvider {
+    public void Use() {}
+}
+
+class OtherFeatureProvider {
+    public void Use() {}
+}
+```
+
+Note that the `SomeFeatureProvider` type and its `Use` method are kept, but the `Use` method will be rewritten to throw.
+
+The simplified model would encourage the above to be rewritten as follows, resulting in the entire type `SomeFeatureProvider` being removed:
+
+```csharp
+
+class FeatureConsumer {
+    static void Run() {
+        if (Features.SomeFeatureIsSupported) {
+            var some = new SomeFeatureProvider();
+            some.Use();
+        }
+        OtherFeatureProvider other = new();
+        other.Use();
+    }
+}
+```
+
+Perhaps we could introduce the simplified model as an optional strict mode for people who are interested in rewriting their code for maximal size savings.
```

Comment:
I think it depends on how popular requirements on type are going to be. I can't think of a similar optimization for requirements on members which so far are much more common.

I agree that this feels more like an "analyzer only" feature, as opposed to "hard requirement for the system".